src/coreclr/src/pal/src/cruntime/wchar.cpp

```diff
@@ -193,47 +194,51 @@ _wcsicmp(
     return ret;
 }
 
-
 /*++
 Function:
-  _wcslwr
-
-Convert a string to lowercase.
-
-Return Value
-
-Returns a pointer to the converted string. Because the modification is
-done in place, the pointer returned is the same as the pointer passed
-as the input argument. No return value is reserved to indicate an
-error.
-
-Parameter
-
-string   Null-terminated string to convert to lowercase
-
-Remarks
+  _wcslwr_s
 
+See MSDN or the man page for _wcslwr_s.
 --*/
-wchar_16 *
+errno_t
 __cdecl
-_wcslwr(
-        wchar_16 *string)
+_wcslwr_s(
+        wchar_16 *string,
+        size_t sz)
 {
-    int i;
-
     PERF_ENTRY(_wcslwr);
-    ENTRY("_wcslwr (string=%p (%S))\n", string?string:W16_NULLSTRING, string?string:W16_NULLSTRING);
+    ENTRY("_wcslwr_s (string=%p (%S))\n", string ? string : W16_NULLSTRING, string ? string : W16_NULLSTRING);
 
-    for (i=0 ; string[i] != 0; i++)
+    size_t fullSize;
+    if (!ClrSafeInt<size_t>::multiply(sz, sizeof(WCHAR), fullSize))
     {
-        string[i] = towlower(string[i]);
+        return 1;
     }
 
-    LOGEXIT("_wcslwr returning wchar_t %p (%S)\n", string?string:W16_NULLSTRING, string?string:W16_NULLSTRING);
-    PERF_EXIT(_wcslwr);
-    return string;
-}
+    WCHAR *copy = (WCHAR *)malloc(fullSize);
```

Comment:
I don't see a reason for making a copy of the string, converting that copy and then copying it back. It seems that the only reason for that was to make the checking mechanisms in `wcscpy_s` kick in instead of putting them into this function. Would you mind modifying this function as follows?
* Move this function to the src/coreclr/src/pal/src/safecrt into a standalone file to follow all the other safecrt files. The validation macros are in src/coreclr/src/pal/src/safecrt/internal_securecrt.h.
* Check that the passed in string is not null and fail if it is. You can use the _VALIDATE_RETURN_ERRCODE macro with EINVAL error code
* Check that the passed in string is null terminated within the specified sz (use PAL_wcsnlen, if it returns sz, fail). You can use the _RETURN_DEST_NOT_NULL_TERMINATED macro to return the failure
* removing using the extra buffer
* optionally, use _FILL_STRING(string, sz,  the_result_of_PAL_wcsnlen+1) to fill the unused part of the buffer in debug builds in the success case.

Reply:
Refactored according to the instructions, thanks! It is interesting that it required adding function prototype in both `mbusafecrt.h` and `pal.h` to satisfy the build which includes palsuite.

Reply:
Thank you! The reason for that is that the prototype in mbusafecrt.h is enclosed in extern "C". Without it, the function would use c++ name mangling.